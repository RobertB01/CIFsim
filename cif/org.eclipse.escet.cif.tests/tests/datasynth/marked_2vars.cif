//////////////////////////////////////////////////////////////////////////////
// Copyright (c) 2010, 2023 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available
// under the terms of the MIT License which is available at
// https://opensource.org/licenses/MIT
//
// SPDX-License-Identifier: MIT
//////////////////////////////////////////////////////////////////////////////

plant p:
  disc int[0..5] x in any, y in any;

  location:
    initial;
    marked (x = 1 and y = 1) or (x = 2 and y = 2);
end
