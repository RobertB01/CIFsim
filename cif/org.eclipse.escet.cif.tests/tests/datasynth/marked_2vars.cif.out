Reading CIF file "datasynth/marked_2vars.cif".
Preprocessing CIF specification (includes checking that the specification is supported).
Converting CIF specification to internal format (BDDs):
    CIF variables and location pointers:
        Nr     Kind               Type       Name  Group  BDD vars  CIF values  BDD values  Values used
        -----  -----------------  ---------  ----  -----  --------  ----------  ----------  -----------
        1      discrete variable  int[0..5]  p.x   0      3 * 2     6 * 2       8 * 2       75%
        2      discrete variable  int[0..5]  p.y   1      3 * 2     6 * 2       8 * 2       75%
        -----  -----------------  ---------  ----  -----  --------  ----------  ----------  -----------
        Total                                      2      12        24          32          75%

    Applying variable ordering:
        Applying 4 orderers, sequentially:
            Applying model variable order:
                Effect: both

            Applying DCSH algorithm:
                Metric: wes
                Relations: legacy
                Effect: var-order
                Number of hyper-edges: 0

                Skipping orderer: no hyper-edges.

            Applying FORCE algorithm:
                Metric: total-span
                Relations: linearized
                Effect: var-order
                Number of hyper-edges: 0
                Maximum number of iterations: 10

                Skipping algorithm: no hyper-edges.

            Applying sliding window algorithm:
                Size: 4
                Metric: total-span
                Relations: linearized
                Effect: var-order
                Number of hyper-edges: 0
                Window length: 2

                Skipping algorithm: no hyper-edges.

    Variable order unchanged.

Starting data-based synthesis.

Showing input and checking for potential problems:
    Invariant (components state plant inv):      true
    Invariant (locations state plant invariant): true
    Invariant (system state plant invariant):    true

    Invariant (components state req invariant):  true
    Invariant (locations state req invariant):   true
    Invariant (system state req invariant):      true

    Initial   (discrete variable 0):             true
    Initial   (discrete variable 1):             true
    Initial   (discrete variables):              true
    Initial   (components init predicate):       true
    Initial   (aut/locs init predicate):         true
    Initial   (auts/locs init predicate):        true
    Initial   (uncontrolled system):             true
    Initial   (system, combined init/plant inv): true
    Initial   (system, combined init/state inv): true

    Marked    (components marker predicate):     true
    Marked    (aut/locs marker predicate):       p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1
    Marked    (auts/locs marker predicate):      p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1
    Marked    (uncontrolled system):             p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1
    Marked    (system, combined mark/plant inv): p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1
    Marked    (system, combined mark/state inv): p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1

    State/event exclusion plants:
        None

    State/event exclusion requirements:
        None

    Uncontrolled system:
        State: (controlled-behavior: ?)

Restricting edge guards to prevent runtime errors:
    No guards changed.

Restricting uncontrolled system behavior using state/event exclusion plant invariants:
    No guards changed.

Initializing edges for being applied.

Restricting uncontrolled system behavior using state plant invariants:
    No restrictions needed.

Initializing controlled behavior:
    Controlled-behavior predicate: true.
    Controlled-initialization predicate: true.

Restricting behavior using state requirements:
    Controlled behavior not changed.

Extending controlled-behavior predicate using variable ranges:
    Controlled behavior: true -> true [range: true, variable: discrete variable "p.x" of type "int[0..5]" (group: 0, domain: 0+1, BDD variables: 3, CIF/BDD values: 6/8)].
    Controlled behavior: true -> true [range: true, variable: discrete variable "p.y" of type "int[0..5]" (group: 1, domain: 2+3, BDD variables: 3, CIF/BDD values: 6/8)].

    Extended controlled-behavior predicate using variable ranges: true.

Restricting behavior using state/event exclusion requirements:
    Guards and controlled behavior not changed.

Restricting behavior using implicit runtime error requirements:
    Controlled behavior not changed.

Re-initializing edges for being applied.

Checking pre-synthesis for events that are never enabled.

Synthesis round 1:
    Computing backward controlled-behavior predicate:
        Backward controlled-behavior: p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1 [marker predicate]

        Controlled behavior: true -> p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1.

    Computing backward uncontrolled bad-state predicate:
        Backward uncontrolled bad-state: (p.x != 2 or p.y != 2) and (p.x != 1 or p.y != 1) [current/previous controlled behavior predicate]

        Controlled behavior not changed.

    Computing forward controlled-behavior predicate:
        Forward controlled-behavior: true [initialization predicate]
        Forward controlled-behavior: true -> p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1 [restricted to current/previous controlled-behavior predicate: p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1]

        Forward controlled-behavior: p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1 [fixed point].

        Controlled behavior not changed.

    Finished: controlled behavior is stable.

Computing final controlled system guards:
    No guards changed.

Cleaning up edges for being applied.

Final synthesis result:
    State: (controlled-behavior: p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1)

Computing initialization predicate of the controlled system.

Controlled system: exactly 2 states.

Determining initialization predicate for output model:
    Initial (synthesis result):            p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1
    Initial (uncontrolled system):         true
    Initial (controlled system):           p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1
    Initial (removed by supervisor):       (p.x != 2 or p.y != 2) and (p.x != 1 or p.y != 1)
    Initial (added by supervisor):         p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1

    Simplifying of controlled system initialization predicate under the assumption of the uncontrolled system initialization predicates:
        Initial: p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1 -> p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1 [assume true].

    Initial (output model):                p.x = 2 and p.y = 2 or p.x = 1 and p.y = 1

Determining supervisor guards for output model.

Checking post-synthesis for events that are never enabled.

Constructing output CIF specification.
Writing output CIF file "datasynth/marked_2vars.ctrlsys.real.cif".
